openapps/csensors/csensors.c

```diff
@@ -0,0 +1,398 @@
+
+\param[in] msg          The received message. CoAP header and options already
+   parsed.
+\param[in] coap_header  The CoAP header contained in the message.
+\param[in] coap_options The CoAP options contained in the message.
+
+\return Whether the response is prepared successfully.
+*/
+owerror_t csensors_receive(
+      OpenQueueEntry_t* msg,
+      coap_header_iht*  coap_header,
+      coap_option_iht*  coap_options
+   ) {
+   owerror_t outcome;
+   uint8_t   id,i;
+   uint32_t  period;
+
+
+   switch (coap_header->Code) {
+      case COAP_CODE_REQ_GET:
+         // reset packet payload
+         msg->payload                     = &(msg->packet[127]);
+         msg->length                      = 0;
+
+         if (coap_options[1].type != COAP_OPTION_NUM_URIPATH) {
+
+            // have CoAP module write links to csensors resources
+            opencoap_writeLinks(msg,COMPONENT_CSENSORS);
+
+            packetfunctions_reserveHeaderSize(msg,1);
+            msg->payload[0]     = COAP_PAYLOAD_MARKER;
+
+            // add return option
+            packetfunctions_reserveHeaderSize(msg,2);
+            msg->payload[0]     = COAP_OPTION_NUM_CONTENTFORMAT << 4 | 1;
+            msg->payload[1]     = COAP_MEDTYPE_APPLINKFORMAT;
+
+         } else {
+
+            for(id=0;id<csensors_vars.numSensors;id++) {
+               if (
+                  memcmp(
+                     coap_options[1].pValue,
+                     csensors_vars.csensors_resource[id].desc.path1val,
+                     csensors_vars.csensors_resource[id].desc.path1len
+                  )==0
+                  ) {
+                  break;
+               }
+            }
+            csensors_fillpayload(msg,id);
+            packetfunctions_reserveHeaderSize(msg,2);
+            msg->payload[0] = (COAP_OPTION_NUM_CONTENTFORMAT << 4) | 1;
+            msg->payload[1] = COAP_MEDTYPE_APPOCTETSTREAM;
+         }
+         // set the CoAP header
+         coap_header->Code                = COAP_CODE_RESP_CONTENT;
+
+         outcome                          = E_SUCCESS;
+
+         break;
+
+      case COAP_CODE_REQ_PUT:
+
+         period = 0;
+
+         for(i=0;i<msg->length;i++) {
+            if ((uint8_t)msg->payload[i]>=(uint8_t)'0' && (uint8_t)msg->payload[i]<=(uint8_t)'0'+9) {
+               period = period*10 + (uint8_t)msg->payload[i] - (uint8_t)'0';
+            } else {
+               period = 0;
+               break;
+            }
+         }
+
+         if (coap_options[1].type == COAP_OPTION_NUM_URIPATH) {
+
+            for(id=0;id<csensors_vars.numSensors;id++) {
+               if (
+                  memcmp(
+                     coap_options[1].pValue,
+                     csensors_vars.csensors_resource[id].desc.path1val,
+                     csensors_vars.csensors_resource[id].desc.path1len
+                  )==0
+                  ) {
+                  break;
+               }
+            }
+
+            csensors_changebehavior(period,id);
```

Comment:
`csensors_changebehavior` is not clear. It looks like you mean `csensors_setPeriod`?